server/tilesource/base.py

```diff
@@ -54,6 +59,104 @@ def __init__(self):
         self.sizeX = None
         self.sizeY = None
 
+    def _calculateWidthHeight(self, width, height, sourceWidth, sourceHeight):
+        """
+        Given a source width and height and a maximum destination width and/or
+        height, calculate a destination width and height that preserves the
+        aspect ratio of the source.
+
+        :param width: the destination width.  None to only use height.
+        :param height: the destination height.  None to only use width.
+        :param sourceWidth: the width of the source data.
+        :param sourceHeight: the height of the source data.
+        :returns: the width and height that is no larger than that specified
+                  and preserves aspect ratio.
+        """
+        if sourceWidth == 0 or sourceHeight == 0:
+            return 0, 0
+        # Constrain the maximum size if both width and height weren't
+        # specified, in case the image is very short or very narrow.
+        if height and not width:
+            width = height * 16
+        if width and not height:
+            height = width * 16
+        if width * sourceHeight > height * sourceWidth:
+            width = max(1, int(sourceWidth * height / sourceHeight))
+        else:
+            height = max(1, int(sourceHeight * width / sourceWidth))
+        return width, height
+
+    def _encodeImage(self, image, encoding='JPEG', jpegQuality=95,
+                     jpegSubsampling=0, **kwargs):
+        """
+        Convert a PIL image into the raw output bytes and a mime type.
+
+        :param image: a PIL image.
+        :param encoding: a valid PIL encoding (typically 'PNG' or 'JPEG').
+                         Must also be in the outputMimeTypes map.
+        :param jpegQuality: the quality to use when encoding a JPEG.
+        :param jpegSubsampling: the subsampling level to use when encoding a
+                                JPEG.
+        """
+        if encoding not in self.outputMimeTypes:
+            raise ValueError('Invalid encoding "%s"' % encoding)
+        if image.width == 0 or image.height == 0:
+            tileData = b''
```

Comment:
Is a 0x0 image really an empty string in both PNG and JPEG? There's no magic number or headers?